src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1355,11 +1411,53 @@ public void InvalidateHttp2Connection(Http2Connection connection)
         {
             lock (SyncObj)
             {
-                if (_http2Connection == connection)
+                Http2Connection[]? localHttp2Connections = _http2Connections;
+
+                if (localHttp2Connections == null)
+                {
+                    return;
+                }
+
+                if (localHttp2Connections.Length == 1)
+                {
+                    // Fast shortcut for the most common case.
+                    if (localHttp2Connections[0] == connection)
+                    {
+                        _http2Connections = null;
+                    }
+                    return;
+                }
+
+                if (TryRemoveHttp2Connection(connection, localHttp2Connections, out Http2Connection[] newHttp2Connections))
+                {
+                    _http2Connections = newHttp2Connections;
+                }
+            }
+        }
+
+        private static bool TryRemoveHttp2Connection(Http2Connection connection, Http2Connection[] localHttp2Connections, out Http2Connection[] newHttp2Connections)
+        {
+            int invalidatedIndex = Array.IndexOf(localHttp2Connections, connection);
```

Comment:
Since this function expects the lock to be held, we should assert that it is held here.

Reply:
I think the lock is not necessary here because it is a **static** function changing only the **locally created array** while the arguments are treated as immutable. Even if `static` is removed, by just looking at this function, it still won't be clear why a lock is taken.

Reply:
Function inlined into the caller.